sympy/solvers/tests/test_solveset.py

```diff
@@ -1087,13 +1106,79 @@ def test_issue_10555():
         ConditionSet(x, Eq(2*f(x) - pi, 0), S.Reals)
 
 
+def test_issue_9824():
+    soln = ImageSet(Lambda(n, 2 * n * pi + pi / 2), S.Integers)
+    assert solveset(sin(x)**2 - 2 * sin(x) + 1, x, S.Reals) == soln
+
+
+def test_issue_9531_and_9606():
+    # complex solution
+    soln = ImageSet(Lambda(n, n * I * pi), S.Integers)
+    assert solveset(sinh(x)) == soln
+    # real solution
+    assert solveset(sinh(x), x, S.Reals) == FiniteSet(0)
+
+
+def test_issue_7914():
+    n = Dummy('n')
+    img1 = ImageSet(Lambda(n, 2*n*pi + pi/6), S.Integers)
+    img2 = ImageSet(Lambda(n, 2*n*pi +5* pi/6), S.Integers)
+    img3 = ImageSet(Lambda(n, 2*n*pi - pi/2), S.Integers)
+    soln = Union(img1, img2, img3, evaluate=False)
+    assert solveset(sin(2 * x) * cos(x) + cos(2 * x) * sin(x) - 1, x) == soln
+
+
+def test_issue_10671():
+    soln = Intersection(Interval(0, pi), ImageSet(Lambda(n, n * pi), S.Integers))
+    assert solveset(sin(y), y, Interval(0, pi)) == soln
+
+
+def test_simplifed_trig_solution():
+    n = Dummy('n', real=True)
+    img1 = ImageSet(Lambda(n, n * pi + pi / 2), S.Integers)
+    img2 = ImageSet(Lambda(n, n * pi / 2 + pi / 6), S.Integers)
+    img3 = ImageSet(Lambda(n, n * pi / 2 + pi / 3), S.Integers)
+    soln = Union(img1, img2, img3)
+    assert solveset(cos(x) + cos(3 * x) + cos(5 * x), x, S.Reals) == soln
```

Comment:
One can read this [gist](https://gist.github.com/Shekharrajak/a5efc840d9a7d3062289f2d9c5f20b16) to understand how the `imageset. _union` is helpful to reduce the number of Imagesets.